### ISSUE TITLE
Overfitting?

### ISSUE THREAD

**Issue by MasIgor** [ultralytics/ultralytics#23752]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi all!

i am having a dataset with about 3000 images. each image has multiple classes on it. (between 1 and 30) 
some classes are rare and appear only 4 or 5 times in the dataset. other classes appear 30 times on the same image and there are multiple images. 

in total i have 360 classes and the quantity is about this:

about 10 classes are rare and have less than 20 instances.
5 classes are common and have 1000-3000 instances each.
the rest are normal, and have about 300 instances each.

i am facing issues with 2 classes in particular, and both have 1000-3000 instances in the dataset. the problem is a bit strange, and is image dependent. on one image every instance is recognized with a score of 0.95. on nother image (same distance, same lighting, same camera, same focus, just a new image..) there are zero recognitions or they are recognized with a score of 0.02. it also happens that the right half of the image is ok, the left part has zero recognitions. but i cannot figure out why. 

the training is done with this command:

yolo detect train model='yolo26l.pt' imgsz=1280 batch=8 epochs=170 data='/mnt/Data/data.yaml' 
augment=True device=[0,1] cache=disk patience=50 fliplr=0

any help is greatly appreciated

edit: here is a sample:

<img width="1251" height="799" alt="Image" src="https://github.com/user-attachments/assets/a488a294-bb67-48fe-8583-1af9d3213fd1" />

<img width="1256" height="846" alt="Image" src="https://github.com/user-attachments/assets/23b6fe7c-748e-4d39-b96f-cea0e3a402be" />

i would expect single crates not to be recognized, not the entire image.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MasIgor, thank you for the detailed report and for using Ultralytics 🚀 This is an automated response to help move things forward quickly; an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
Given your report includes image-dependent behavior, please also share an MRE if possible so we can reliably reproduce what you’re seeing 🔍

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
This doesn’t look like classic “overfitting” so much as a data/augmentation or labeling inconsistency: since you’ve explicitly disabled horizontal flips with `fliplr=0`, the model can easily learn a left/right spatial prior and then miss objects on the “unseen” side—can you try a short retrain with `fliplr` left at default (or set to a nonzero value) and see if the left-half failures disappear (more on symptoms in the [overfitting glossary](https://www.ultralytics.com/glossary/overfitting) and general checks in [Troubleshooting common YOLO issues](https://docs.ultralytics.com/guides/yolo-common-issues/))?

Also, please paste `yolo checks` output and attach one “bad” image *plus its exact matching label file* (the `.txt` in your `labels/` dir); if you edit any labels, delete any `*.cache` files under `labels/` before re-running so you’re not validating against stale cached labels. To sanity-check whether the model is producing low-confidence boxes that are just being filtered out, run predict at a very low `conf` and confirm you’re using the same `imgsz` as training:
```bash
yolo checks
yolo detect predict model=path/to/best.pt source=path/to/bad.jpg imgsz=1280 conf=0.001 save_txt=True
```

**Comment by Y-T-G** [comment]
If you trained on similar looking images, then the model overfit. Images need to be diverse. Not look similar, like closely spaced frames of videos.

**Comment by MasIgor** [comment]
thanks all for you rhelp.. i checked the overfitting glossary as sugested, and tried a smaller model as suggested in there. went from L to S and it solved all problems. now it works great.. didnt expect the S to be powerful enoough to learn 320 classes that sometimes are very similar.

**Comment by MasIgor** [comment]
ok, i tought it to be solved, but after some hours it reappered.. here is a sample photo that is OK:


this one is not ok:


they have been taken on the very same 2 cameras, a few minutes apart. one image found nothing. (the things you see are other detectors) and one image found things only in the right part of the image. this happens on one image every 30 more or less..

any idea on what it could be?

**Comment by MasIgor** [comment]
just to mention, i did follow your instructions, and tried to do inf. with a very low conf and exported it to txt.. it finds a few things, but all very very low scores.. like 0.1 or 0.05 or things like thtat..

**Comment by Y-T-G** [comment]
What's your training command?

**Comment by MasIgor** [comment]
> What's your training command?

`yolo detect train model='yolo26s.pt' imgsz=1280 batch=26 epochs=250 data='/mnt/Data/data.yaml' augment=True device=[0,1] cache=disk patience=50 fliplr=0.1 flipud=0.1 degrees=1.0 scale=0.1`

**Comment by Y-T-G** [comment]
You can try training YOLO11s

**Comment by MasIgor** [comment]
thanks will try that today. what is the difference with that? why do you expect it to work better?

**Comment by glenn-jocher** [comment]
`yolo11s.pt` is the previous-generation Ultralytics YOLO architecture, so the main reason to try it here is as a clean A/B sanity check: if **YOLO11s is consistently OK on the same “bad” frames where YOLO26s collapses**, it strongly points to a YOLO26 training/augmentation/hyperparam interaction; if **both** fail on the same frames, it’s almost always a data/label/domain issue rather than “model size.” For context on what true overfitting typically looks like, see the [overfitting glossary](https://www.ultralytics.com/glossary/overfitting).

If you test it, keep everything else identical so the comparison is meaningful:

```bash
yolo detect train model=yolo11s.pt data=/mnt/Data/data.yaml imgsz=1280 batch=26 epochs=250 device=[0,1] cache=disk patience=50 fliplr=0.1 flipud=0.1 degrees=1.0 scale=0.1
```

After training, please paste `yolo checks` and attach one failing image *with its exact matching* label `.txt` so we can verify there isn’t an image-specific label/size/EXIF inconsistency driving the “one in ~30 frames” behavior.

**Comment by MasIgor** [comment]
thanks, will do that today. for reference, i added a few of the failing frames to my dataset and trained again overnight, and now those that failed yesterday did not fail anymore today. but i cannot figure out what could possibly be different on those 30 frames.. i will revert back the dataset, delete the cache, and try training a y11s network today. will report back tomorrow

**Comment by MasIgor** [comment]
i made the test with yolo11s and the results are the same. 

this is the content of the txt file:

20 0.150331 0.635261 0.067561 0.245334
20 0.215857 0.634463 0.0668504 0.244156
20 0.902891 0.635509 0.0626942 0.234878
20 0.965068 0.636008 0.0648045 0.235302
20 0.841441 0.636764 0.0623926 0.238174
20 0.154958 0.379797 0.0668006 0.239873
20 0.279995 0.634107 0.0645308 0.242952
20 0.996672 0.641959 0.00665512 0.227436
91 0.0684914 0.884395 0.102558 0.231209



but since adding these images to the dataset solved the problem i guess the problem could be something like the distance of the object, so that it gets slightly out of focus..

**Comment by glenn-jocher** [comment]
Since **YOLO26s and YOLO11s fail on the same frames** and the problem disappears after you add those “bad” frames to training, this points to **missing domain coverage** (often slight defocus/motion blur, compression artifacts, vignetting/uneven illumination, or subtle camera pipeline changes) rather than “overfitting”; a good next step is to *measure* how different the failing frames are (blur is easy to quantify) and then deliberately collect/label more of those conditions and/or loosen your augmentation instead of keeping it very conservative (see the [Ultralytics data augmentation guide](https://www.ultralytics.com/blog/the-ultimate-guide-to-data-augmentation-in-2025)).  

```python
import cv2, numpy as np
from pathlib import Path

def blur_score(p):
    img = cv2.imread(str(p), cv2.IMREAD_GRAYSCALE)
    if img is None:
        return None
    return cv2.Laplacian(img, cv2.CV_64F).var()  # lower = blurrier

good = Path("good.jpg")
bad  = Path("bad.jpg")
print("good blur score:", blur_score(good))
print("bad  blur score:", blur_score(bad))
```

**Comment by MasIgor** [comment]
thank you very much, will make these tests and try to identify what it is missing.. i wasnt able to find defocussing int he augmentation process.. am i missing it or isnt there one?

**Comment by glenn-jocher** [comment]
There isn’t a dedicated “defocus” knob in the built-in YOLO hyperparams, but you can approximate it via the Albumentations blur transforms (e.g. `Blur`, `MedianBlur`) that Ultralytics will apply automatically during training once Albumentations is installed; see the [Albumentations integration docs](https://docs.ultralytics.com/integrations/albumentations/) and the [Albumentations augmentation walkthrough](https://www.ultralytics.com/blog/using-albumentations-augmentations-to-diversify-your-data).

```bash
pip install albumentations
```

If you need *stronger/more frequent* defocus-style blur than the defaults, the next step is a small code tweak to the Albumentations transform list in `ultralytics/data/augment.py` (editable install / fork) since it’s not currently exposed as a CLI training arg.

**Comment by MasIgor** [comment]
Thank you very much for your support! will try it. Closing this issue

**Comment by glenn-jocher** [comment]
Glad this unblocked you—if the behavior returns, please open a fresh issue with one failing image + its matching label `.txt` and `yolo checks` output, following our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can reproduce it quickly.
